go-controller/pkg/ovn/egressgw.go

```diff
@@ -564,61 +566,68 @@ func (oc *Controller) addGWRoutesForPod(gateways []*gatewayInfo, podIfAddrs []*n
 	return nil
 }
 
-// deletePerPodGRSNAT removes per pod SNAT rules that are applied to the GR where the pod resides if
-// there are no gateways
-func deletePerPodGRSNAT(nbClient libovsdbclient.Client, node string, podIPNets []*net.IPNet) error {
-	gr := util.GetGatewayRouterFromNode(node)
-	nats := make([]*nbdb.NAT, 0, len(podIPNets))
-	var nat *nbdb.NAT
+// buildPerPodGRSNAT builds per pod SNAT rules towards the nodeIP that are applied to the GR where the pod resides
+// if allSNATs flag is set, then all the SNATs (including against egressIPs if any) for that pod will be returned
+func buildPerPodGRSNAT(nodeName string, watchFactory *factory.WatchFactory, podIPNets []*net.IPNet, allSNATs bool) ([]*nbdb.NAT, error) {
 	var err error
-	for _, podIPNet := range podIPNets {
-		podIP := podIPNet.IP.String()
-		mask := GetIPFullMask(podIP)
-		_, fullMaskPodNet, err := net.ParseCIDR(podIP + mask)
-		if err != nil {
-			klog.Errorf("Invalid IP: %s and mask: %s combination, error: %v", podIP, mask, err)
-			continue
-		}
-		nat = libovsdbops.BuildRouterSNAT(nil, fullMaskPodNet, "", nil)
-		nats = append(nats, nat)
-	}
-	err = libovsdbops.DeleteNatsFromRouter(nbClient, gr, nats...)
-	if err != nil {
-		return fmt.Errorf("failed to delete SNAT rule for pod on gateway router %s, "+
-			"error: %v", gr, err)
-	}
-	return nil
-}
-
-func addPerPodGRSNAT(nbClient libovsdbclient.Client, watchFactory *factory.WatchFactory, pod *kapi.Pod, podIfAddrs []*net.IPNet) error {
-	nodeName := pod.Spec.NodeName
 	node, err := watchFactory.GetNode(nodeName)
 	if err != nil {
-		return fmt.Errorf("failed to get node %s: %v", nodeName, err)
+		return nil, fmt.Errorf("failed to get node %s: %v", nodeName, err)
 	}
 	l3GWConfig, err := util.ParseNodeL3GatewayAnnotation(node)
 	if err != nil {
-		return fmt.Errorf("unable to parse node L3 gw annotation: %v", err)
+		return nil, fmt.Errorf("unable to parse node L3 gw annotation: %v", err)
 	}
-	nats := make([]*nbdb.NAT, 0, len(l3GWConfig.IPAddresses)*len(podIfAddrs))
+	nats := make([]*nbdb.NAT, 0, len(l3GWConfig.IPAddresses)*len(podIPNets))
 	var nat *nbdb.NAT
-	gr := types.GWRouterPrefix + nodeName
 	for _, gwIPNet := range l3GWConfig.IPAddresses {
 		gwIP := gwIPNet.IP.String()
-		for _, podIPNet := range podIfAddrs {
+		for _, podIPNet := range podIPNets {
 			podIP := podIPNet.IP.String()
 			if utilnet.IsIPv6String(gwIP) != utilnet.IsIPv6String(podIP) {
 				continue
 			}
 			mask := GetIPFullMask(podIP)
 			_, fullMaskPodNet, err := net.ParseCIDR(podIP + mask)
 			if err != nil {
-				return fmt.Errorf("invalid IP: %s and mask: %s combination, error: %v", podIP, mask, err)
+				return nil, fmt.Errorf("invalid IP: %s and mask: %s combination, error: %v", podIP, mask, err)
+			}
+			if allSNATs {
+				nat = libovsdbops.BuildRouterSNAT(nil, fullMaskPodNet, "", nil)
```

Comment:
@tssurya is it possible that two different podIPNets (from different pods) would evaluate to the same fullMaskPodNet and that we'd add duplicate NAT entries now?

Reply:
I don't think addition/updates would be a problem, since addOrUpdateNatToRouterOps() de-dupes; but would deletion be a problem?

eg if two different pods create the same BuildRouterSNAT() NAT object, but of course addOrUpdateNatToRouterOps() de-dupes and only one entry is added to the router, but then one of the pods gets deleted (but the other still exists) then that one pod's NAT entry would match the single one on the router in DeleteNatsFromRouterOps() and delete it, even though the other pod still exists?

Maybe I'm over-thinking it though.

Reply:
hmm, not sure I follow, 

when you say 2 different podIPNets from different pods - at a given time we cannot have 2 pods in the cluster with the same IP right?

`net.ParseCIDR(podIP + mask)` is just adding the `podIP+/32`or `podIP+/128`, so I'd expect `podIP` to be always unique?

so if that pod gets deleted then only NATs belonging to that pod should get deleted. Example:
```
sh-5.1# ovn-nbctl lr-nat-list GR_ovn-worker2
TYPE             EXTERNAL_IP        EXTERNAL_PORT    LOGICAL_IP            EXTERNAL_MAC         LOGICAL_PORT
snat             172.18.0.6                          10.244.2.4
snat             172.18.0.3                          10.244.2.4
```

where 172.18.0.3 is nodeIP and 172.18.0.6 is externalIP.

Reply:
I am not sure how two different pods can end up with same SNAT